pkg/tbtc/signing.go

```diff
@@ -2,378 +2,372 @@ package tbtc
+					signature: nil,
+					endBlock:  0,
+					err:       err,
+				}
+
+				return
+			}
+
+			// Do not cancel the loopCtx upon function exit immediately and
+			// continue to broadcast signing done checks until the successful
+			// attempt timeout. This way we maximize the chance that other
+			// members, especially the ones not participating in the successful
+			// signature attempt, receive the done checks as well.
+			go func() {
+				defer cancelLoopCtx()
+
+				err := se.waitForBlockFn(
+					loopCtx,
+					loopResult.attemptTimeoutBlock,
+				)
+				if err != nil {
+					signingLogger.Warnf(
+						"[member:%v] failed waiting for signing "+
+							"loop stop signal: [%v]",
+						signer.signingGroupMemberIndex,
+						err,
+					)
+				}
+			}()
+
+			signingLogger.Infof(
+				"[member:%v] generated signature [%v] at block [%v]",
+				signer.signingGroupMemberIndex,
+				loopResult.result.Signature,
+				loopResult.latestEndBlock,
+			)
+
+			signerOutcomesChan <- &signerOutcome{
+				signature: loopResult.result.Signature,
+				endBlock:  loopResult.latestEndBlock,
+				err:       nil,
+			}
+		}(currentSigner)
 	}
 
-	// Make sure we always use just the smallest required count of
-	// signing members for performance reasons
-	if len(includedMembersIndexes) > srl.chainConfig.HonestThreshold {
-		// #nosec G404 (insecure random number source (rand))
-		// Shuffling does not require secure randomness.
-		rng := rand.New(rand.NewSource(
-			srl.attemptSeed + int64(srl.attemptCounter),
-		))
-		// Sort in ascending order just in case.
-		sort.Slice(includedMembersIndexes, func(i, j int) bool {
-			return includedMembersIndexes[i] < includedMembersIndexes[j]
-		})
-		// Shuffle the included members slice to randomize the
-		// selection of additionally excluded members.
-		rng.Shuffle(len(includedMembersIndexes), func(i, j int) {
-			includedMembersIndexes[i], includedMembersIndexes[j] =
-				includedMembersIndexes[j], includedMembersIndexes[i]
-		})
-		// Get the surplus of included members and add them to
-		// the excluded members list.
-		excludedMembersIndexes = append(
-			excludedMembersIndexes,
-			includedMembersIndexes[srl.chainConfig.HonestThreshold:]...,
-		)
-		// Sort the resulting excluded members list in ascending order.
-		sort.Slice(excludedMembersIndexes, func(i, j int) bool {
-			return excludedMembersIndexes[i] < excludedMembersIndexes[j]
-		})
+	signerOutcomes := make([]*signerOutcome, 0)
+
+	for {
+		select {
+		case outcome := <-signerOutcomesChan:
+			signerOutcomes = append(signerOutcomes, outcome)
+
+			if len(signerOutcomes) == len(se.signers) {
+				outcomes := slices.CompactFunc(
+					signerOutcomes,
+					func(o1, o2 *signerOutcome) bool {
+						return o1.signature.Equals(o2.signature) &&
+							o1.endBlock == o2.endBlock &&
+							reflect.DeepEqual(o1.err, o2.err)
+					},
+				)
+
+				if len(outcomes) != 1 {
+					return nil, 0, fmt.Errorf("signers came to different outcomes")
```

Comment:
We are losing the context about what the outcomes were, no? 

Why do we care whether they were the same or not? We could introduce a custom error type aggregating multiple potential errors. If at least one of the signers failed, we return that error and we do not care about anything else.

Reply:
Here we basically want to establish a single outcome for all signers we control. If they all came to the same outcome, we are good. If they diverged somehow, we must fail. All signers create multiple log messages during their work and upon completion so I think there is no point to display the errors again here, especially given the fact that the aggregate will be very long and poorly readable.

Reply:
I do not think it is the right choice to fail if they diverged. Imagine the following scenario: we have a wallet controlled by one operator. 50 of the signing group members claim they completed the signature and 1 other claims it failed because of some error that happened after the signing done check. The signature was completed, t+1 members had to agree on signing it and successfully execute the signing. I think we should use that signature. This example can be extended to the case of two operators with 1 signing group member failing on each operator or even more operators. This would be even more dangerous given the fraud proofs and the fact the signature was computed successfully. 

I think the current contract of `signing.Execute` is that it completes successfully only if all signing group members confirmed they are done by sending `signingDoneMessage` and all signatures match. If so, I think we are fine with writing to `signerOutcomeChan` only for the first successful execution and just log errors returned from `retryLoop.start` like we already do. The first result that arrives into `signingDoneMessage` is returned.

I think this will make the code a lot simpler, we just need to clearly express this constraint in `signing.Execute` documentation and maybe mention it again in a comment to the code when returning the first result from `signerOutcomeChan`.

Reply:
Yeah, that makes sense! See 1f244c4a